ui/src/components/Shell/index.jsx

```diff
@@ -70,33 +68,200 @@ export default class Shell extends Component {
 
       // Create a shell client, with interface similar to child_process
       // With an additional method client.resize(cols, rows) for TTY sizing.
-      if (version === '1') {
-        this.client = new DockerExecClient(options);
-        await this.client.execute();
+      switch (version) {
+        // docker worker
+        case '1':
+          io.sendString = d => this.client && this.client.stdin.write(d);
+          io.onVTKeystroke = io.sendString;
 
-        // Wrap client.resize to switch argument ordering
-        const { resize } = this.client;
+          this.client = new DockerExecClient(options);
+          await this.client.execute();
 
-        this.client.resize = (c, r) => resize.call(this.client, r, c);
-      } else {
-        io.writeUTF8(`inteactive shell API version ${version} not supported`);
+          this.client.resize.apply(this.client, [
+            terminal.screenSize.height,
+            terminal.screenSize.width,
+          ]);
+
+          this.client.on('exit', code => {
+            io.writeUTF8(`\r\nRemote shell exited: ${code}\r\n`);
+            terminal.uninstallKeyboard();
+            terminal.setCursorVisible(false);
+          });
+
+          this.client.resize(
+            terminal.screenSize.width,
+            terminal.screenSize.height
+          );
+          io.onTerminalResize = (c, r) => this.client.resize(c, r);
+          this.client.stdout.on('data', data =>
+            io.writeUTF8(DECODER.decode(data))
+          );
+          this.client.stderr.on('data', data =>
+            io.writeUTF8(DECODER.decode(data))
+          );
+          this.client.stdout.resume();
+          this.client.stderr.resume();
+
+          break;
+        // generic worker
+        case '2':
+          this.wsClient = new WebSocket(url);
+          this.cmd = [];
+          this.currentCmd = 0;
+          this.prompt = '$ ';
+          // minCol is the minimum column to start printing at
+          // and to stop deleting at
+          // accounts for the prompt "$ "
+          this.minCol = 2;
+
+          io.sendString = d => {
+            switch (d) {
+              // return key
+              case '\r': {
+                if (!this.cmd[this.currentCmd]) {
+                  io.print(`\n\r${this.prompt}`);
+
+                  break;
+                }
+
+                io.println('');
+
+                this.wsClient &&
+                  this.wsClient.send(`${this.cmd[this.currentCmd]}\n`);
```

Comment:
why it can be not defined?
```suggestion
                this.wsClient?.send(`${this.cmd[this.currentCmd]}\n`);
```

Reply:
I guess it _probably shouldn't_ be undefined, but I just used a similar approach to what they have for docker worker implementation.